emails/tests/mgmt_process_emails_from_sqs_tests.py

```diff
@@ -284,62 +361,124 @@ def test_no_body_deleted(mock_sqs_client, caplog, test_settings):
-    mock_sns_inbound_logic.side_effect = (internal_error, None)
+def test_ses_python_error(
+    mock_sns_inbound_logic: Mock,
+    mock_sqs_client: Mock,
+    test_settings: SettingsWrapper,
+    caplog: LogCaptureFixture,
+) -> None:
+    """The command catches processing failures."""
+    test_settings.PROCESS_EMAIL_MAX_SECONDS = 2
+    mock_sns_inbound_logic.side_effect = ValueError("bad stuff")
     msg = fake_sqs_message(json.dumps(TEST_SNS_MESSAGE))
     mock_sqs_client.return_value = fake_queue([msg], [])
     call_command(COMMAND_NAME)
     summary = summary_from_exit_log(caplog)
     assert summary["total_messages"] == 1
     assert summary["failed_messages"] == 1
     msg.delete.assert_not_called()
+    rec2 = caplog.records[1]
+    assert rec2.msg == "Message processed"
+    rec2_extra = log_extra(rec2)
+    assert rec2_extra["success"] is False
+    assert rec2_extra["error_type"] == "ValueError"
+    assert rec2_extra["error"] == "bad stuff"
+
+
+def test_ses_slow(
+    mock_sns_inbound_logic: Mock,
+    mock_sqs_client: Mock,
+    mock_process_pool_future: Mock,
+    test_settings: SettingsWrapper,
+    caplog: LogCaptureFixture,
+) -> None:
+    test_settings.PROCESS_EMAIL_MAX_SECONDS_PER_MESSAGE = 120
+    msg = fake_sqs_message(json.dumps(TEST_SNS_MESSAGE))
+    mock_sqs_client.return_value = fake_queue([msg], [], [])
+    mock_process_pool_future._is_stalled.side_effect = [False, False, True]
+    call_command(COMMAND_NAME)
+    summary = summary_from_exit_log(caplog)
+    assert summary["total_messages"] == 1
+    msg.delete.assert_called()
+    rec2 = caplog.records[1]
+    assert rec2.msg == "Message processed"
+    rec2_extra = log_extra(rec2)
+    assert rec2_extra["success"] is True
+    assert rec2_extra["message_process_time_s"] < 120.0
+    assert rec2_extra["subprocess_setup_time_s"] == 1.0
+    assert mock_process_pool_future._timeouts == [1.0]
+
+
+def test_ses_timeout(
+    mock_sns_inbound_logic: Mock,
+    mock_sqs_client: Mock,
+    mock_process_pool_future: Mock,
+    test_settings: SettingsWrapper,
+    caplog: LogCaptureFixture,
+) -> None:
+    test_settings.PROCESS_EMAIL_MAX_SECONDS_PER_MESSAGE = 120
+    msg = fake_sqs_message(json.dumps(TEST_SNS_MESSAGE))
+    mock_sqs_client.return_value = fake_queue([msg], [], [])
+    mock_process_pool_future._is_stalled.return_value = True
+    call_command(COMMAND_NAME)
+    summary = summary_from_exit_log(caplog)
+    assert summary["total_messages"] == 1
+    assert summary["failed_messages"] == 1
+    msg.delete.assert_not_called()
+    rec2 = caplog.records[1]
+    assert rec2.msg == "Message processed"
+    rec2_extra = log_extra(rec2)
+    assert rec2_extra["success"] is False
+    assert rec2_extra["error"] == "Timed out after 120.0 seconds."
+    assert rec2_extra["message_process_time_s"] >= 120.0
+    assert rec2_extra["subprocess_setup_time_s"] == 1.0
+    assert mock_process_pool_future._timeouts == [1.0] * 60
```

Comment:
question (non-blocking): Again I don't understand this `_timeouts` property: why is it a list of 60 `1.0` values when the message timed out after 120 seconds? That seems like there should be 120 `1.0` values?

Reply:
Good question! The loop includes a call to `time.monotonic`, which also increments the mocked clock, so each loops is 2 fake seconds. However, it makes more sense if `future.wait` does _not_ increment the clock. I'll make that change.